runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -39,3 +39,27 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
    TR_ASSERT_FATAL(false, "This helper implements platform specific code for Fieldwatch, which is currently not supported on ARM64 platforms.\n");
    }
 	
+TR::Register *J9::ARM64::TreeEvaluator::arraylengthEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(cg->fe());
+   // ldrw R, [B + contiguousSize]        
+   // cmp R, 0                            
+   // cselw R, [B + discontiguousSize]    
+   //
+   TR::Register *objectReg = cg->evaluate(node->getFirstChild());
+   TR::Register *lengthReg = cg->allocateRegister();
+   TR::Register *condReg = cg->allocateRegister(TR_CCR);
+
+   TR::MemoryReference *contiguousArraySizeMR = new (cg->trHeapMemory()) TR::MemoryReference(objectReg, fej9->getOffsetOfContiguousArraySizeField(), 4, cg);
+   TR::MemoryReference *discontiguousArraySizeMR = new (cg->trHeapMemory()) TR::MemoryReference(objectReg, fej9->getOffsetOfDiscontiguousArraySizeField(), 4, cg);
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, lengthReg, contiguousArraySizeMR);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::ccmpimmw, node, condReg, lengthReg, 0);
+   generateTrg1MemInstruction(cg, TR::InstOpCode::cselw, node, lengthReg, discontiguousArraySizeMR);
```

Comment:
Please read the specification for the CSEL instruction (section C6.2.61 in the ARM Architecture Reference Manual for ARMv8-A) for its format and semantics.

You'll find that this instruction actually operates on 3 registers and a condition:
```
   CSEL <Xd>, <Xn>, <Xm>, <cond>
```
and who's operation is defined (very loosely) as:
```
   Xd = cond ? Xn : Xm
```
This means you'll have to load the discontiguous array length into a register first.

Looking into this deeper, I don't believe we have an instruction class for this kind of instruction yet, so you'll have to create one in OMR.  I think creating a `ARM64CondTrg1Src2Instruction` that extends `ARM64Trg1SrcInstruction` is the way to go.  One of the constructor APIs for such an instruction should look like this:
```
   ARM64CondTrg1Src2Instruction(
         TR::InstOpCode::Mnemonic op,
         TR::Node *node,
         TR::ARM64ConditionCode cc,
         TR::Register *treg,
         TR::Register *s1reg,
         TR::Register *s2reg,
         TR::CodeGenerator *cg) :
      ARM64Trg1Src2Instruction(op, node, treg, s1reg, s2reg, cg),
         _conditionCode(cc)
      {
      ...
      }
   ```
Please have a look at creating a new instruction kind.  You can probably follow the pattern of other instructions.

@knn-k in case he has any different opinions on this.

Reply:
I agree with @0xdaryl on making a new Instruction class based on existing `ARM64Trg1Src2Instruction` plus `ARM64ConditionCode`.
Please look at `ARM64Trg1Src2ShiftedInstruction` for how to add an extra element to `ARM64Trg1Src2Instruction`.